pkg/runner/runner.go

```diff
@@ -0,0 +1,238 @@
+	Execute(request executionworkertypes.ExecuteRequest) (*executionworkertypes.ExecuteResult, error)
+	Pause(id string) error
+	Resume(id string) error
+	Abort(id string) error
+}
+
+type runner struct {
+	worker               executionworkertypes.Worker
+	outputRepository     testworkflow.OutputRepository
+	executionsRepository testworkflow.Repository
+	configRepository     configRepo.Repository
+	emitter              *event.Emitter
+	metrics              metrics.Metrics
+	dashboardURI         string
+	storageSkipVerify    bool
+
+	watching sync.Map
+}
+
+func New(
+	worker executionworkertypes.Worker,
+	outputRepository testworkflow.OutputRepository,
+	executionsRepository testworkflow.Repository,
+	configRepository configRepo.Repository,
+	emitter *event.Emitter,
+	metrics metrics.Metrics,
+	dashboardURI string,
+	storageSkipVerify bool,
+) Runner {
+	return &runner{
+		worker:               worker,
+		outputRepository:     outputRepository,
+		executionsRepository: executionsRepository,
+		configRepository:     configRepository,
+		emitter:              emitter,
+		metrics:              metrics,
+		dashboardURI:         dashboardURI,
+		storageSkipVerify:    storageSkipVerify,
+	}
+}
+
+func (r *runner) monitor(ctx context.Context, execution testkube.TestWorkflowExecution) error {
+	defer r.watching.Delete(execution.Id)
+
+	var notifications executionworkertypes.NotificationsWatcher
+	for i := 0; i < 10; i++ {
+		notifications = r.worker.Notifications(ctx, execution.Id, executionworkertypes.NotificationsOptions{})
+		if notifications.Err() == nil {
+			break
+		}
+		if errors.Is(notifications.Err(), registry.ErrResourceNotFound) {
+			// TODO: should it mark as job was aborted then?
+			return registry.ErrResourceNotFound
+		}
+		time.Sleep(500 * time.Millisecond)
+	}
+	if notifications.Err() != nil {
+		return errors.Wrapf(notifications.Err(), "failed to listen for '%s' execution notifications", execution.Id)
+	}
+
+	logs, err := NewExecutionLogsWriter(r.outputRepository, execution.Id, execution.Workflow.Name, r.storageSkipVerify)
+	if err != nil {
+		return err
+	}
+	saver, err := NewExecutionSaver(ctx, r.executionsRepository, execution.Id, logs)
+	if err != nil {
+		return err
+	}
+	defer logs.Cleanup()
+
+	currentRef := ""
+	var lastResult *testkube.TestWorkflowResult
+	for n := range notifications.Channel() {
+		if n.Temporary {
+			continue
+		}
+
+		if n.Output != nil {
+			saver.AppendOutput(*n.Output)
+		} else if n.Result != nil {
+			lastResult = n.Result
+			// Don't send final result until everything is finished
+			if n.Result.IsFinished() {
+				continue
+			}
+			saver.UpdateResult(*n.Result)
+		} else {
+			if n.Ref != currentRef && n.Ref != "" {
+				currentRef = n.Ref
+				err = logs.WriteStart(n.Ref)
+				if err != nil {
+					// FIXME: what to do then?
+					panic("logs write ref error")
+				}
+			}
+			_, err = logs.Write([]byte(n.Log))
+			if err != nil {
+				// FIXME: what to do then?
+				panic("logs write error")
```

Comment:
should we panic there, might be logging errors?

Reply:
It shouldn't rather ever happen, so I have added an error log + continue (to handle it gracefully)